### ISSUE TITLE
firebase.messaging.getToken() edgecase when push notification is set back to default (ask)

### ISSUE THREAD

**Issue by thomasmburke** [firebase/firebase-js-sdk#2364]
### [REQUIRED] Describe your environment

  * Operating System version: macOS Mojave Version 10.14.6
  * Browser version: Google Chrome Version 78.0.3904.97
  * Firebase SDK version: 7.0.0
  * Firebase Product: Firebase Cloud Messaging

### [REQUIRED] Describe the problem
The issue I have identified was introduced in version 7.0.0 of the firebase-js-sdk. I do not run into this issue in versions previous to 7.0.0. The issue occurs when a use grants notification permissions, gets a FCM token, then updates the notification permissions in the browser back to Ask (default), reloads the page, grants notification permissions again in Chrome, and get an error when trying to getToken at this point. Note: subsequent calls to getToken don't face this issue.

#### Steps to reproduce:

The problem occurs after taking the following steps:
- Call Notification.requestPermission() and user grants permissions
- Call firebase.messaging.getToken()
- Set notification permission in browser back to Ask (default)
![image](https://user-images.githubusercontent.com/40719837/68976050-c8536f00-07a9-11ea-926d-d81af6eea498.png)

- Call Notification.requestPermission() and user grants permissions
- call firebase.messaging.getToken()
- get DELETE  https://fcmregistrations.googleapis.com/v1/projects/{project_id}/regis.../{FCM_token} 500 error and this specific error from FCM: "FirebaseError: Messaging: A problem occured while unsubscribing the user from FCM: FirebaseError: Messaging: A problem occured while unsubscribing the user from FCM: Internal error encountered. (messaging/token-unsubscribe-failed). (messaging/token-unsubscribe-failed)."
- call messaging.getToken() again
- get new token, no error this time.

Others have encountered this issue as well on Google Groups:
https://groups.google.com/a/google.com/forum/#!msg/firebase-developer-cuf/dGU1SnJ1cz0/GMa_WvWGBwAJ

#### Relevant Code:

The firebase javascript messaging quickstart encounters this issue:
https://github.com/firebase/quickstart-js/tree/master/messaging

Specifically, I think the update needs to be made here:
firebase-js-sdk/packages/messaging/src/controllers/base-controller.ts
Lines: 136-137

**Comment by bjgrosse** [comment]
I also ran into this error just as the OP described.

**Comment by mindiii006** [comment]
I am also getting this issue.

**Comment by rommelpe** [comment]
Thank for the report. I was able to reproduce this using the [quick start guide](https://github.com/firebase/quickstart-js/tree/master/messaging) and filed an internal report (b/144697421).

![error-messaging](https://user-images.githubusercontent.com/56452310/69078065-0387c380-0a72-11ea-93db-ead765e45ef5.jpg)

I'm just wondering if you y'all happen to consistently facing this issue. There are times that I can only reproduce it on the initial load or upon requesting permission several times. Are there any particular patterns you notice that significantly contributes to the issue (aside from what mentioned)?

**Comment by thomasmburke** [comment]
Hi rommelpe,

This issue only arises after the a reset of notifications permissions and reload of page and then works properly with an additional reloaded pages. Aside from what I mentioned there are no other patterns that contribute to the issue identified.

Thanks, and let me know if there are any other questions.

**Comment by Nox04** [comment]
I'm getting the same issue when I remove the push notifications permission and reload.

**Comment by gestewa** [comment]
I am also getting this issue.

**Comment by thayannevls** [comment]
I'm getting the same error :(

**Comment by mmermerkaya** [comment]
Thanks for the report. I can consistently reproduce the behavior and I'm investigating the issue.

I'd like to ask everyone to please refrain from posting "This is happening to me too" comments and instead use the add reaction feature on the OP to add a thumbs up to issues that you consider important.

**Comment by TrackMyToken** [comment]
Any update on this issue?

**Comment by mmermerkaya** [comment]
Hello, sorry for the lack of updates. Looks like the problem is in our backend, but there wasn't much progress as we've had a code freeze last week (because of black friday). I should be able to give you a better update or even a fix later this week.

**Comment by NaghamHalabi** [comment]
Can we have the latest updates regarding this issue?

**Comment by mmermerkaya** [comment]
Hey everyone, our backend team is still investigating this. Since it's taking longer than my initial estimate, I just submitted a PR to temporarily suppress this error.

**Comment by spawnrider** [comment]
Also reproduced this issue and need a fix
Using Firebase version 7.5.2

**Comment by kdluu** [comment]
This is happening to me too

**Comment by hsubox76** [comment]
The temporary change to suppress the error, mentioned by @mmermerkaya above, is in 7.6.0 which has just been released.

**Comment by CodeByBryan** [comment]
Using 7.6.0 and am still having the same issue!

**Comment by mmermerkaya** [comment]
If you're only seeing the error in the console, that's normal. My PR is still logging the error in the console, but it should not stop the `getToken` call from completing successfully.

**Comment by CodeByBryan** [comment]
I am receiving a token value when https://firebaseinstallations.googleapis.com/v1/projects/.../authTokens:generate is called, but get a "500" error when calling getToken()

**Comment by GoktuqCan** [comment]
I get this error when using "127.0.0.1" instead of localhost.
BUT if i use "localhost" even the getToken() is not working. Showing CORS policy error.

- On localhost error is(cannot getToken())
```
OPTIONS https://firebaseinstallations.googleapis.com/v1/projects/{project_id}/installations/{some_token}/authTokens:generate 403

Access to fetch at 'https://firebaseinstallations.googleapis.com/v1/projects/{project_id}/installations/{some_token}/authTokens:generate' from origin 'http://localhost:3000' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
```

- On 127.0.0.1 error(getting token but shows this error) is:
```
DELETE https://fcmregistrations.googleapis.com/v1/projects/{project_id}/regis.../{FCM_token} 500

FirebaseError: Messaging: A problem occured while unsubscribing the user from FCM: Internal error encountered. (messaging/token-unsubscribe-failed). (messaging/token-unsubscribe-failed).
```

**Comment by mmermerkaya** [comment]
The CORS error looks like an unrelated issue. Can you create a new bug report for that?

**Comment by Ridermansb** [comment]
@mmermerkaya did you created new issue for CORS  problem?
I'm having the same issue.

**Comment by ctavan** [comment]
@Ridermansb @mmermerkaya @GoktuqCan @CodeByBryan could you check if #2467 matches the issue that you are observing?

**Comment by CodeByBryan** [comment]
I am having the issue in a Chrome Extension. when getToken() is called I recieve:

> DELETE https://fcmregistrations.googleapis.com/v1/projects/{project_id}/regis.../{FCM_token} 500
FirebaseError: Messaging: A problem occured while unsubscribing the user from FCM: Internal error encountered. (messaging/token-unsubscribe-failed). 
 (messaging/token-unsubscribe-failed).<

I am not getting any CORS errors

**Comment by niklat22** [comment]
Same error for user resubscribe:

```
base-controller.ts:245 FirebaseError: Messaging: A problem occured while unsubscribing the user from FCM: FirebaseError: Messaging: A problem occured while unsubscribing the user from FCM: Internal error encountered. (messaging/token-unsubscribe-failed). (messaging/token-unsubscribe-failed).
    at Ik.<anonymous> (https://www.gstatic.com/firebasejs/7.6.0/firebase.js:1:811146)
    at https://www.gstatic.com/firebasejs/7.6.0/firebase.js:1:52093
    at Object.next (https://www.gstatic.com/firebasejs/7.6.0/firebase.js:1:52198)
    at n (https://www.gstatic.com/firebasejs/7.6.0/firebase.js:1:50968)
```

**Comment by Depender-Sethi** [comment]
Waiting for the fix, for the time being I'm using a way around by checking the error code and calling the function again because this error is thrown only the first time.

```
const turnOnNotification = async () => {
    const uid = firebase.auth().currentUser.uid;
    const messaging = fb.messaging();
    try {
      await messaging.requestPermission();
      const token = await messaging.getToken();
      // send to server
    } catch (err) {
      if (err.code === "messaging/token-unsubscribe-failed")
        turnOnNotification();
    } 
  };
```

**Comment by Marc-Tanne** [comment]
> Hello, sorry for the lack of updates. Looks like the problem is in our backend, but there wasn't much progress as we've had a code freeze last week (because of black friday). I should be able to give you a better update or even a fix later this week.

are there any new updates on this. Our company is getting held up by this issue. we are using `"firebase": "7.6.1"` and the error is not suppressed in this version. we are calling get token like this: `messaging.getToken("SENDER ID", "FCM");`

**Comment by kdastan** [comment]
I tried on version 7.9.1

If I grant permission to show notification, the token generates just fine, but if I remove permission from chrome settings and then try to reload the page and after grant permission again in a console I'm having error like 
```
FirebaseError: Messaging: A problem occurred while unsubscribing the user from FCM: FirebaseError: Messaging: A problem occurred while unsubscribing the user from FCM: Requested entity was not found. (messaging/token-unsubscribe-failed). (messaging/token-unsubscribe-failed).
``` 
but at the same time, a new valid token generates.

**Comment by rommelpe** [comment]
Sorry to hear about your issue, @kdastan. Other dev seems to encounter it, but I can't get the same behavior. Can I ask you to share the [following](https://github.com/firebase/firebase-js-sdk/issues/2643#issuecomment-588120333)? Thanks.

**Comment by srameshr** [comment]
@rommelpe I am on 7.11
```
  <script src="https://www.gstatic.com/firebasejs/7.11.0/firebase-app.js"></script>
  <script src="https://www.gstatic.com/firebasejs/7.11.0/firebase-messaging.js"></script>
```
And this issue occurs for getToken() fails with bad request 400.

```
https://fcmregistrations.googleapis.com/v1/projects/frills-e9796/registrations
```

**Comment by srameshr** [comment]
@akuales Go to Application -> Service workers and unregister your service worker.

**Comment by Jeybee89** [comment]
The same problem here. 

`main.js:2852 Unable to get permission to notify. FirebaseError: Messaging: A problem occured while subscribing the user to FCM: Requests to this API fcmregistrations.googleapis.com method google.firebase.fcm.registration.v1.RegistrationApi.CreateRegistration are blocked. (messaging/token-subscribe-failed).` 

trying everything and still getting this error

**Comment by haimlvov10** [comment]
> The same problem here.
> 
> `main.js:2852 Unable to get permission to notify. FirebaseError: Messaging: A problem occured while subscribing the user to FCM: Requests to this API fcmregistrations.googleapis.com method google.firebase.fcm.registration.v1.RegistrationApi.CreateRegistration are blocked. (messaging/token-subscribe-failed).`
> 
> trying everything and still getting this error

Same here.

**Comment by Jeybee89** [comment]
I think I fixed this problem by changing credentials in the google cloud console (API restrictions)

**Comment by michielmulders** [comment]
@Jeybee89 Can you expand on this how you did it? The problem still persists.

**Comment by Jeybee89** [comment]
@michielmulders  do to google cloud console (https://console.cloud.google.com/), select your project then APIs & Services > Credentials > API keys > Browser Key > API restrictions > Don't restrict key

**Comment by manemmao** [comment]
@Jeybee89 Error repeated with this config too. version 7.14.1.

not causing any special problems. just ugly 404 error in console

**Comment by Zelears** [comment]
Same problem here.

**Comment by Zelears** [comment]
> Waiting for the fix, for the time being I'm using a way around by checking the error code and calling the function again because this error is thrown only the first time.
> 
> ```
> const turnOnNotification = async () => {
>     const uid = firebase.auth().currentUser.uid;
>     const messaging = fb.messaging();
>     try {
>       await messaging.requestPermission();
>       const token = await messaging.getToken();
>       // send to server
>     } catch (err) {
>       if (err.code === "messaging/token-unsubscribe-failed")
>         turnOnNotification();
>     } 
>   };
> ```

It fixed <3 
Only minus is that you sending every time existing token to server

**Comment by daovangiangbk** [comment]
In my case, `getToken` successfully registrations after 404 error. But i always received the same token, even i already unregistered Service worker and asking for permission again. And received token is totally useless ( FCM api return success while nothing happen at browser)

**Comment by Mwaseemzakir** [comment]
![issue](https://user-images.githubusercontent.com/44539744/85369496-a479b800-b546-11ea-8db0-6e7a060a6f65.PNG)

**Comment by aforty** [comment]
Enabling "FCM Registration API" for my google api key fixed the issue for me.

![Screen Shot 2020-07-17 at 13 07 13](https://user-images.githubusercontent.com/523683/87812609-767a5180-c82e-11ea-9079-932d2f07aa91.png)

**Comment by Stas-Buzunko** [comment]
updating to latest version (firebase@7.21.1) fixed for me

**Comment by YeetOrBeYate** [comment]
@Stas-Buzunko firebase 7.21.1? like the package or in your service worker?

**Comment by jonathanmv** [comment]
I'm using `7.21.1` in both the package.json and the service worker. I was using the key from another project by mistake. Check that the same is not happening to you 😂

**Comment by Symasia** [comment]
For future visitors: Goto Dev Tools -> Application -> Clear storage.

I feel like this issue only takes place when you disable notifications and enable them again. The fact that the service worker is registered again can also be a culprit in some form, but just unregistering the service worker did not solve the problem for me. My firebase version is 7.9.1, so I guess the issue still persists but it does not affect anything critically.

**Comment by jeffykle** [comment]
> Enabling "FCM Registration API" for my google api key fixed that particular issue for me.

My key was already unrestricted and I'm experiencing the issue.

**Comment by Kamran92** [comment]
I have the same problem 8.2.7
![image](https://user-images.githubusercontent.com/51258443/108628311-4b627b00-746b-11eb-9825-33c1f13d3a03.png)

**Comment by dr-aiuta** [comment]
Actually, investigating here the error that @Kamran92 appointed, it seems that token is still valid, although the message says the "entity was not found". It returns success when sending a test message from my server.

**Comment by dr-aiuta** [comment]
> do you use Vue ? @dr-aiuta

No

**Comment by arnav-kr** [comment]
same issue here

**Comment by sergdenisov** [comment]
Have the same issue (404) with the `8.4.1` version. Found 2 ways to reproduce (2-nd step):

1. Everything is OK with an app registration and getting a token.
2. Block notifications, then allow notifications (without reloading a page) **OR** unregister your service worker in browser settings.
3. Reload a page.

<img width="1152" alt="Screenshot 2021-04-20 at 18 49 38" src="https://user-images.githubusercontent.com/5359222/115427897-87ffe800-a20a-11eb-83f9-78c34ff05e89.png">

**Comment by choubsovann** [comment]
![image](https://user-images.githubusercontent.com/24521478/116340174-e92a4b80-a808-11eb-8e5e-a9af38df4ebd.png)
Please help me!

**Comment by marciocamello** [comment]
The same error here

**Comment by 10eputzen** [comment]
It is horrible to always delete the Cache once this happens ...

**Comment by unfinishedCode** [comment]
I tried some of the solutions listed on here and on a [Stck Overflow question](https://stackoverflow.com/questions/58325574/firebase-fcm-javascript-error-when-request-permission-error-status-500) regarding this bug, it didnt seem to do the trick. The one solution on here causes an endless loop. The other solution is to use a lower version, which is blocked by a redirect. To solve this 'bug' I enabled 5 APIs in [Google's Cloud Console](https://console.cloud.google.com/), and restricted my push notification key to the following APIs. The five APIs that I restricted the key to are:

- FCM Registration API
- Firebase Cloud Messaging API
- Firebase Installations API
- Firebase Management API
- Firebase Remote Config API

I also included a picture. I hope this helps somebody. This bug doesn't impact functionality, but its very unprofessional to see an error in the console. 
<img width="567" alt="Screen Shot 2021-06-22 at 10 03 37 AM" src="https://user-images.githubusercontent.com/4739272/122939537-b51f6280-d341-11eb-99b6-254a239e39d7.png">

**Comment by sahejsingh** [comment]
I ran into this issue this morning. 
```
POST https://fcmregistrations.googleapis.com/v1/projects/my-project-id/registrations 401 (Unauthorized)
<call-stack>
index.html:630 Unable to get permission. FirebaseError: Messaging: A problem occurred while subscribing the user to FCM: Request is missing required authentication credential. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project. (messaging/token-subscribe-failed).
    at api.ts:67
    at tslib.es6.js:102
    at Object.next (tslib.es6.js:83)
    at r (tslib.es6.js:73)

```
here is how it happened
- created `client-1`, got notifications working.
- for some reason I happened to delete the `client-1` from firebase console.
- created `client-2`
- copied the new `firebaseConfig` from console.
- got the error now.
- Noticed, the `Web API Key` is the same for both the clients. there are not restrictions on the web api key.
- I have followed everything I could (unregister service worker)

Firebase JS Version: 8.6.8

scripts imported like so:
```
<script src="https://www.gstatic.com/firebasejs/8.6.8/firebase-app.js"></script>
<script src="https://www.gstatic.com/firebasejs/8.6.8/firebase-messaging.js"></script>
var firebaseConfig = {
  // CONFIG COPIED FROM FIREBASE CONSOLE.
};

firebase.initializeApp(firebaseConfig);

const messaging = firebase.messaging();


  messaging
    .getToken({ vapidKey: "MY_PUB_KEY" })
    .then((currentToken) => {
      console.log(currentToken)
    })
    .catch(function (err) {
      // handle error
    });
```

**Comment by lamquochoang** [comment]
> I ran into this issue this morning.
> 
> ```
> POST https://fcmregistrations.googleapis.com/v1/projects/my-project-id/registrations 401 (Unauthorized)
> <call-stack>
> index.html:630 Unable to get permission. FirebaseError: Messaging: A problem occurred while subscribing the user to FCM: Request is missing required authentication credential. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project. (messaging/token-subscribe-failed).
>     at api.ts:67
>     at tslib.es6.js:102
>     at Object.next (tslib.es6.js:83)
>     at r (tslib.es6.js:73)
> ```
> 
> here is how it happened
> 
> * created `client-1`, got notifications working.
> * for some reason I happened to delete the `client-1` from firebase console.
> * created `client-2`
> * copied the new `firebaseConfig` from console.
> * got the error now.
> * Noticed, the `Web API Key` is the same for both the clients. there are not restrictions on the web api key.
> * I have followed everything I could (unregister service worker)
> 
> Firebase JS Version: 8.6.8
> 
> scripts imported like so:
> 
> ```
> <script src="https://www.gstatic.com/firebasejs/8.6.8/firebase-app.js"></script>
> <script src="https://www.gstatic.com/firebasejs/8.6.8/firebase-messaging.js"></script>
> var firebaseConfig = {
>   // CONFIG COPIED FROM FIREBASE CONSOLE.
> };
> 
> firebase.initializeApp(firebaseConfig);
> 
> const messaging = firebase.messaging();
> 
> 
>   messaging
>     .getToken({ vapidKey: "MY_PUB_KEY" })
>     .then((currentToken) => {
>       console.log(currentToken)
>     })
>     .catch(function (err) {
>       // handle error
>     });
> ```

I also have the same problem since noon today, I have tried many ways but still can't fix it.

**Comment by omkarajagunde** [comment]
> > I ran into this issue this morning.
> > ```
> > POST https://fcmregistrations.googleapis.com/v1/projects/my-project-id/registrations 401 (Unauthorized)
> > <call-stack>
> > index.html:630 Unable to get permission. FirebaseError: Messaging: A problem occurred while subscribing the user to FCM: Request is missing required authentication credential. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project. (messaging/token-subscribe-failed).
> >     at api.ts:67
> >     at tslib.es6.js:102
> >     at Object.next (tslib.es6.js:83)
> >     at r (tslib.es6.js:73)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > here is how it happened
> > 
> > * created `client-1`, got notifications working.
> > * for some reason I happened to delete the `client-1` from firebase console.
> > * created `client-2`
> > * copied the new `firebaseConfig` from console.
> > * got the error now.
> > * Noticed, the `Web API Key` is the same for both the clients. there are not restrictions on the web api key.
> > * I have followed everything I could (unregister service worker)
> > 
> > Firebase JS Version: 8.6.8
> > scripts imported like so:
> > ```
> > <script src="https://www.gstatic.com/firebasejs/8.6.8/firebase-app.js"></script>
> > <script src="https://www.gstatic.com/firebasejs/8.6.8/firebase-messaging.js"></script>
> > var firebaseConfig = {
> >   // CONFIG COPIED FROM FIREBASE CONSOLE.
> > };
> > 
> > firebase.initializeApp(firebaseConfig);
> > 
> > const messaging = firebase.messaging();
> > 
> > 
> >   messaging
> >     .getToken({ vapidKey: "MY_PUB_KEY" })
> >     .then((currentToken) => {
> >       console.log(currentToken)
> >     })
> >     .catch(function (err) {
> >       // handle error
> >     });
> > ```
> 
> I also have the same problem since noon today, I have tried many ways but still can't fix it.

Yup facing this issue since morning

**Comment by jamesbrian0323** [comment]
> I ran into this issue this morning.
> 
> ```
> POST https://fcmregistrations.googleapis.com/v1/projects/my-project-id/registrations 401 (Unauthorized)
> <call-stack>
> index.html:630 Unable to get permission. FirebaseError: Messaging: A problem occurred while subscribing the user to FCM: Request is missing required authentication credential. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project. (messaging/token-subscribe-failed).
>     at api.ts:67
>     at tslib.es6.js:102
>     at Object.next (tslib.es6.js:83)
>     at r (tslib.es6.js:73)
> ```
> 
> here is how it happened
> 
> * created `client-1`, got notifications working.
> * for some reason I happened to delete the `client-1` from firebase console.
> * created `client-2`
> * copied the new `firebaseConfig` from console.
> * got the error now.
> * Noticed, the `Web API Key` is the same for both the clients. there are not restrictions on the web api key.
> * I have followed everything I could (unregister service worker)
> 
> Firebase JS Version: 8.6.8
> 
> scripts imported like so:
> 
> ```
> <script src="https://www.gstatic.com/firebasejs/8.6.8/firebase-app.js"></script>
> <script src="https://www.gstatic.com/firebasejs/8.6.8/firebase-messaging.js"></script>
> var firebaseConfig = {
>   // CONFIG COPIED FROM FIREBASE CONSOLE.
> };
> 
> firebase.initializeApp(firebaseConfig);
> 
> const messaging = firebase.messaging();
> 
> 
>   messaging
>     .getToken({ vapidKey: "MY_PUB_KEY" })
>     .then((currentToken) => {
>       console.log(currentToken)
>     })
>     .catch(function (err) {
>       // handle error
>     });
> ```

I also have the same issue - getting 401 error.
I tried to delete the project and create it again, but still not working.

**Comment by rachelrab** [comment]
This also happens to me, started happening a day ago - not consistent.

**Comment by deivisondfs2** [comment]
This also happens to me :(
Yesterday it was normal and today I'm getting this error:
![image](https://user-images.githubusercontent.com/16122029/124115101-c48a6400-da43-11eb-8e5c-2c9e7751f93a.png)

**Comment by deivisondfs2** [comment]
Does anyone have any updates?
Or where can we try to get a status?
I just created a new account and entered the new credentials but the error remains.

**Comment by gach87** [comment]
I'm experiencing this problem also, created a new project and still remains.

**Comment by deivisondfs2** [comment]
Ah this is the same issue:
firebase/firebase-js-sdk#5081
Let's centralize the discussion over there.

**Comment by ChromeQ** [comment]
If it is the same issue then why go over there? That was raised yesterday,
this is an established issue with many comments over a year.

Close that one and centralise over here

**Comment by sahejsingh** [comment]
I do have to say, I just realised that that my issue was specific to getting the token.
The error was: `messaging/token-subscribe-failed`, http code=`401`.
whereas this issue is about `unsubscribe` and http code=`500`.

**Comment by atzzCokeK** [comment]
> For future visitors: Goto Dev Tools -> Application -> Clear storage.
> 
> I feel like this issue only takes place when you disable notifications and enable them again. The fact that the service worker is registered again can also be a culprit in some form, but just unregistering the service worker did not solve the problem for me. My firebase version is 7.9.1, so I guess the issue still persists but it does not affect anything critically.

I tried this solution. It is fixed, but when you change the status for notification, it happens again.
thanks @Symasia

**Comment by jahooma** [comment]
Hi y'all,

We've had this problem (or related ones) for a long time. I think I finally got it all working today. This is how.

I believe Firebase messaging is just buggy when you try to register your own service worker instead of providing the firebase-messaging-sw.js file.

So the fix was to create that file and then reference at the end of my own `service-worker.js`:

`importScripts('firebase-messaging-sw.js')`

Then you no longer need to pass serviceWorkerRegistration into getToken, simply call:

`const token = await messaging.getToken({ vapidKey: config.publicVapidKey })`

And it just works!

**Comment by Ashatanchik** [comment]
I'm having exactly the same issue

`FirebaseError: Messaging: A problem occurred while unsubscribing the user from FCM`

Using `"firebase": "8.3.2"`

It seems like this issue takes place due to cache erros. So to fix it I am always clearing cache, that's pity :(

Unfortunately I don't have an ability to use firebase-messaging-sw.js, that's why @jahooma method wont work for me

**Comment by Mwamiq381** [comment]
Have the same issue (404) with the 8.3.1 please guide me how can i fix this issu
![error](https://user-images.githubusercontent.com/92384050/138139003-b478bd4b-ea7c-4ba8-9de0-d38e56ca30a1.PNG)
e

**Comment by rockmandash** [comment]
I'm using `firebase-messaging-sw.js` and still can reproduce this bug using @sergdenisov 's two method, so I think @jahooma 's method is not working.

**Comment by mohammad-kazemzadeh** [comment]
This issue still persists and only way to "temporarily" solve it is to `clear storage`. 
`firebase v9.6.7`

**Comment by Gubancs** [comment]
I get this error also with firebase v9.6.10 in Chrome but not in Firefox!

_FirebaseError: Messaging: A problem occurred while unsubscribing the user from FCM: FirebaseError: Messaging: A problem occurred while unsubscribing the user from FCM: Requested entity was not found. (messaging/token-unsubscribe-failed). (messaging/token-unsubscribe-failed).
    at index.esm2017.js:428:29
    at Generator.next (<anonymous>)
    at asyncGeneratorStep (asyncToGenerator.js:3:1)
    at _next (asyncToGenerator.js:25:1)
    at _ZoneDelegate.invoke (zone.js:372:1)
    at Object.onInvoke (core.mjs:25568:33)
    at _ZoneDelegate.invoke (zone.js:371:1)
    at Zone.run (zone.js:134:1)
    at zone.js:1275:1
    at _ZoneDelegate.invokeTask (zone.js:406:1)_

I can confirm the clearing the application storage on the developer console solves the problem temporarily, but  it is not the best solution.

**Comment by jorgeroncero** [comment]
This is happening to me aswell with version Firebase JS SDK 9.9.0 wrapped within Flutter Firebase SDK. 

It happens really often while testing PWA and browser navigation. It seems like the Firebase SDK is returning an already deleted token no matter what. Even if in the code there is a call to get a new token, it's returning the same invalid token.

**Comment by jaysonmulwa** [comment]
Same issue

**Comment by jaysonmulwa** [comment]
This worked for me [https://github.com/firebase/firebase-js-sdk/issues/2643#issuecomment-596817538](url)

**Comment by srihari9n** [comment]
@jaysonmulwa  the comment you posted link to seems not be there anymore. Can you tell what was the solution mentioned in it?

**Comment by jaysonmulwa** [comment]
Hi @srihari9n , It's been quite a while, will try to recall what the solution was.

**Comment by jswilliams** [comment]
I'm seeing this randomly as well. When this happened back in 2021 it ended up being a backend issue on Google's side that they fixed. See: https://github.com/firebase/firebase-js-sdk/issues/5081 

I'm wondering if there was a regression and it's happening again?

@jaysonmulwa if you can remember the work around we'd be able to use it as well.

**Comment by huynhquangkieu** [comment]
> Have the same issue (404) with the `8.4.1` version. Found 2 ways to reproduce (2-nd step):
> 
> 1. Everything is OK with an app registration and getting a token.
> 2. Block notifications, then allow notifications (without reloading a page) **OR** unregister your service worker in browser settings.
> 3. Reload a page.
> 
> <img alt="Screenshot 2021-04-20 at 18 49 38" width="1152" src="https://user-images.githubusercontent.com/5359222/115427897-87ffe800-a20a-11eb-83f9-78c34ff05e89.png">

It still work for me in 2023 ✌️

**Comment by Plinpod** [comment]
Seeing the same issues on `9.17.1` as well

**Comment by amm834** [comment]
this error is still producing in the firebase version  `^9.22.2` and also it should be fixed.

**Comment by nickorsk2017** [comment]
Have it

**Comment by nickorsk2017** [comment]
Not actual problem.

Just save and check token in database.

**Comment by boazblake** [comment]
this is an annoying bug.

**Comment by zarqani** [comment]
the issue is produced in the Firebase version ^10.1.0

**Comment by sahanenes** [comment]
An error occurred while retrieving token: DOMException: Failed to execute 'subscribe' on 'PushManager': The provided applicationServerKey is not valid.  
for chrome , i got this error after notification permission . 
"firebase": "^10.1.0",

**Comment by aditivin** [comment]
Still able to reproduce. Why is it taking so long to fix this?

**Comment by windycloud** [comment]
I encountered this error about a month ago, and it was really frustrating. After multiple debugging attempts, I was able to fix it, and I haven't encountered it since. Although I'm not certain about the root cause of this error, I suspect it was triggered by changes in the idToken. Changing the idToken seemed to lead to the issue of unregistering the FCM token. After reviewing the places where getIdTokenResult and getIdToken were used, especially when setting forceRefresh to true, which caused a change in the idToken, the error no longer occurred. I don't recall the details, but the problem seemed to be related to modifying the idToken. I hope this information can help someone else eliminate a similar issue.

**Comment by anmolraina25** [comment]
When we update the site permissions from browser settings (while our website is not active), for example, I changed the notification permission to Blocked from Allow, and then back to Allow, now I'd expect to start receiving push notifications (right?). But the fcm v1 api returns 'Requested entity not found'. I believe it's a similar case as resetting the permissions as @thomasmburke has mentioned.

In any case in the source code, the error has been quietly handled. Is there any update here, team firebase?

**Comment by galdames** [comment]
Any news?

**Comment by hsubox76** [comment]
There have been many backend changes and 5 major version updates since this issue was originally opened and the details we need to address the current issue may be getting lost in the long history of comments where many of these don't apply anymore.

Anyone experiencing this problem on the latest version, feel free to start a fresh issue where we can focus on the current problem in the current version of the SDK. Provide a full description of what you are currently experiencing and the steps to reproduce it consistently.
